src/serializers/HdfSerializer.cpp

```diff
@@ -0,0 +1,209 @@
+
+}
+
+
+bool HdfSerializer::ready() {
+	//todo: check whether the file exists
+	return true;
+}
+
+void HdfSerializer::writeHeader() {
+	const H5std_string  FILE_NAME(hdf_filename);
+	file = H5::H5File(FILE_NAME, H5F_ACC_TRUNC);
+
+}
+
+
+void HdfSerializer::write(const IfcGeom::BRepElement<real_t>* o) {
+
+	std::string guid = o->guid();
+
+	H5::Group elementGroup;
+
+	H5::Group meshGroup;
+	H5::DataSet positionsDataset;
+	H5::DataSet normalsDataset;
+	H5::DataSet indicesDataset;
+
+	H5::Group OCCTGroup;
+	H5::DataSet OCCTDataset;
+	
+	const IfcGeom::Representation::BRep& brepmesh = o->geometry();
+	const IfcGeom::Representation::Serialization *serialization = new IfcGeom::Representation::Serialization(brepmesh);
+	std::string brep_data = serialization->brep_data();
+
+	const IfcGeom::TriangulationElement<real_t>*triangular_element = new IfcGeom::TriangulationElement<real_t>(*o);
+	const IfcGeom::Representation::Triangulation<real_t>& mesh = triangular_element->geometry();
+	const int vcount = (int)mesh.verts().size() / 3;
+	const int fcount = (int)mesh.faces().size() / 3;
+	const bool isyup = settings().get(SerializerSettings::USE_Y_UP);
+
+	std::string value = o->type();
+
+	if (fcount > 0) {
+
+		guids.insert(guid);
+		elementGroup = file.createGroup(guid);
+		meshGroup = elementGroup.createGroup("Triangle Mesh");
+
+		OCCTGroup = elementGroup.createGroup("OCCT Data");
+
+		H5::StrType str_type(0, H5T_VARIABLE);
+		H5:: DataSpace attrdspace(H5S_SCALAR);
+		H5::Attribute att = elementGroup.createAttribute("IFC entity type", str_type, attrdspace);
+		att.write(str_type, value);
+
+		const int   RANK = 2;
+		hsize_t     dimsf[2];
+		dimsf[0] = vcount;
+		dimsf[1] = 1;
+		H5::DataSpace dataspace(RANK, dimsf);
+
+		hsize_t     dimsfaces[2];
+		dimsfaces[0] = fcount;
+		dimsfaces[1] = 1;
+		H5::DataSpace face_dataspace(RANK, dimsfaces);
+
+		const int RANK_OCCT = 1;
+		hsize_t     dimsocct[2];
+		dimsocct[0] = 1;
+		dimsfaces[1] = 1;
+		H5::DataSpace occt_dataspace(RANK_OCCT, dimsocct);
+		OCCTDataset = OCCTGroup.createDataSet(DATASET_NAME_OCCT, str_type, occt_dataspace);
+		OCCTDataset.write(brep_data, str_type);
+
+		indicesDataset = meshGroup.createDataSet(DATASET_NAME_INDICES, mtype1, face_dataspace);
+		positionsDataset = meshGroup.createDataSet(DATASET_NAME_POSITIONS, mtype2, dataspace);
+		normalsDataset = meshGroup.createDataSet(DATASET_NAME_NORMALS, mtype3, dataspace);
+
+		for (std::vector<real_t>::const_iterator it = mesh.verts().begin(); it != mesh.verts().end(); ) {
+			const real_t x = *(it++);
+			const real_t y = *(it++);
+			const real_t z = *(it++);
+
+			if (isyup) {
+				double_data_container.push_back(x);
+				double_data_container.push_back(z);
+				double_data_container.push_back(-y);
+
+			} else {
+				double_data_container.push_back(x);
+				double_data_container.push_back(y);
+				double_data_container.push_back(z);
+
+			}
+
+		
+		}
+		positionsDataset.write(double_data_container.data(), mtype2);
```

Comment:
In theory you cannot do this. You cannot feed HDF5 a vector<x>. While you constructed a compound datatype that mirrors `s1_t`. If you want to do this it needs to be a `std::vector<s1_t>`. Because you do not have the guarantee that the compiler chooses to layout those things equivalently in memory. Coincidently there is a good chance it will, because a float is rather big, but it could well be that a compiler adds some padding after `s1_t.c` for ex.

My recommendation is that you make it (vcount, 3) dimension dataset with just simple floats. Same for indices.

Reply:
That way all of the struct/compound logic can be removed (a shame I know, but we can save the complex datatypes for the voxels probably :))